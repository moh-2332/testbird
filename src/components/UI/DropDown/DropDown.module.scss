.dropdown {
    position: relative;
    color: #333;
    cursor: default;
  }
  
  .dropdown .close {
    position: absolute;
    right: 7px;
    top: 12px;
    cursor: pointer;
  }
  
  .dropdown .close.open {
    border-color: transparent transparent #999;
    border-width: 0 5px 5px;
  }
  
  .dropdown .selected-value input {
    line-height: 1.5;
    font-size: 1rem;
    background-color: #fff;
    box-shadow: 0 1px 8px rgb(0 0 0 / 20%);
    border-radius: 2px;
    box-sizing: border-box;
    cursor: default;
    outline: none;
    padding: 8px 52px 8px 10px;
    transition: all 200ms ease;
    width: 100%;
    border: none;
  }
  
  .dropdown .options {
    background-color: #fff;
    box-shadow: 0 1px 8px rgb(0 0 0 / 20%);
    box-sizing: border-box;
    margin-top: -1px;
    max-height: 200px;
    overflow-y: auto;
    position: absolute;
    top: 100%;
    width: 100%;
    z-index: 1000;
  }
  
  .dropdown .option {
    box-sizing: border-box;
    color: rgba(51, 51, 51, 0.8);
    cursor: pointer;
    display: block;
    padding: 8px 10px;
  }
  
  .dropdown .option.selected,
  .dropdown .option:hover {
    background-color: var(--secondary-color);
    color: #333;
  }