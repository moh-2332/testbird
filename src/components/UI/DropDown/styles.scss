@import '../../../styles/colors';
@import '../../../styles/dimentions';

.dropdown {
  position: relative;
  cursor: default;
  min-width: 280px;

  &__control {

    &__selected-value {
      padding: $large $medium;
      position: relative;
      box-sizing: border-box;
      background-color: $light;
      box-shadow: 0 1px 8px rgb(0 0 0 / 20%);


      input {
        font-size: 1rem;
        background-color: #fff;
        box-sizing: border-box;
        cursor: default;
        outline: none;
        transition: all 200ms ease;
        width: 100%;
        border: none;
        border-bottom: 2px solid $primary;
        padding-bottom: $small;
      } 
    }

    &__close {
      position: absolute;
      right: 20px;
      top: 20px;
      cursor: pointer;
    }
  }
  
  &__options {
    background-color: $light;
    box-shadow: 0 1px 8px rgb(0 0 0 / 20%);
    box-sizing: border-box;
    margin-top: -1px;
    max-height: 200px;
    overflow-y: auto;
    position: absolute;
    top: 100%;
    width: 100%;
    z-index: 1000;
    
    &--option {
      box-sizing: border-box;
      cursor: pointer;
      display: block;
      padding: $small $medium;
  
      &:hover {
        background-color: $secondary;
      }
    }
  }

  &__not-found {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: $large 0;

    &--title {
      color: $primary;
      font-size: 1.1rem;
    }

    &--description {
      font-size: 0.725rem;
    }
  }
}