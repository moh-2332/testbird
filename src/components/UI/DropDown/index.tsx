import React, { useState } from 'react';

import "./styles.scss"
import Icon from '../Icon';

interface DropDownProps {
    options: Array<any>,
    id: string,
    label: string,
    avatar?: string,
    value: string,
    notFoundFilter?: {
        title: string,
        description: string
    }
    onOptionSelected: (value: any) => any,
    onClose: () => void,
}

const DropDown: React.FC<DropDownProps> = ({ options, id, label, avatar, value, notFoundFilter, onOptionSelected, onClose }) => {
    const [searchKey, setSearchKey] = useState("");

    const optionSelectedHandler = (option: any) => {
        onOptionSelected(option)
        setSearchKey("");
        onClose();
    }

    const filter = (options: Array<any>) => {
        return options.filter(option => option[label].toLowerCase().indexOf(searchKey.toLowerCase()) > -1) || [];
    }

    return (
        <div className="dropdown" data-testid="dropdown">
            <div className="dropdown__control">
                <div className="dropdown__control__selected-value">
                    <input
                        value={searchKey}
                        placeholder={value}
                        onChange={(e) => setSearchKey(e.target.value)} />
                </div>
                <div className="dropdown__control__close" onClick={onClose}><span className="material-icons md-12 remove">close</span></div>
            </div>
            <div className="dropdown__options">
                {filter(options).length ? filter(options).map(option => {
                    return (
                        <div
                            key={option[id]}
                            className="dropdown__options--option"
                            onClick={(e) => optionSelectedHandler(option)}>
                            {avatar && <Icon><img src={`avatars/${option[avatar]}`} /></Icon>}
                            <div className="dropdown__options--option__label">{option[label]}</div>
                        </div>
                    )
                }) : <div className="dropdown__not-found" data-testid="not-found">
                        <span className="dropdown__not-found--title">{notFoundFilter?.title}</span>
                        <span className="dropdown__not-found--description">{notFoundFilter?.description}</span>
                    </div>
                }
            </div>
        </div>
    );
}

export default DropDown;