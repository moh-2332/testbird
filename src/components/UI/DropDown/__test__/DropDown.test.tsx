import React from "react";

import { screen, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import DropDown from "../";

const handler = jest.fn();
const options = [
    {
        username: "username1",
        role: "role1",
        picture: "avatar-default.png",
        id: 1
    }
]

describe("DropDown tests", () => {
    it("Should match the snapshot", () => {
        const dropDownList = renderer.create(<DropDown options={options} id="id" label="username" value="select option" onOptionSelected={handler} onClose={handler} />);
        expect(dropDownList).toMatchSnapshot();
    });

    it("Should renders correctly", () => {
        const { container } = render(<DropDown options={options} id="id" label="username" value="select option" onOptionSelected={handler} onClose={handler} />);

        const optionsList = container.getElementsByClassName("dropdown__options--option");
        expect(optionsList.length).toBe(1);

        const textBox = screen.getByRole("textbox");
        expect(textBox).toBeInTheDocument();
    });

    it("Should renders the not-found status correctly", () => {
        render(<DropDown options={options} id="id" label="username" value="select option" onOptionSelected={handler} onClose={handler} />);

        const textBox = screen.getByRole("textbox");
        userEvent.type(textBox, "aaaa");

        const notFoundOption = screen.getByTestId("not-found");
        expect(notFoundOption).toBeInTheDocument();

    });

    it("Should run the onclick function after selecting the option", () => {
        const { container } = render(<DropDown options={options} id="id" label="username" value="select option" onOptionSelected={handler} onClose={handler} />);

        const optionsList = container.getElementsByClassName("dropdown__options--option");
        userEvent.click(optionsList[0]);

        expect(handler).toBeCalled();
    });
});