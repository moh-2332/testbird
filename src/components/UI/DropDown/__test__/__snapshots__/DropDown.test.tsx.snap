// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown tests Should match the snapshot 1`] = `
<div
  className="dropdown"
  data-testid="dropdown"
>
  <div
    className="dropdown__control"
  >
    <div
      className="dropdown__control__selected-value"
    >
      <input
        onChange={[Function]}
        placeholder="select option"
        value=""
      />
    </div>
    <div
      className="dropdown__control__close"
      onClick={[MockFunction]}
    >
      <span
        className="material-icons md-12 remove"
      >
        close
      </span>
    </div>
  </div>
  <div
    className="dropdown__options"
  >
    <div
      className="dropdown__options--option"
      onClick={[Function]}
    >
      <div
        className="dropdown__options--option__label"
      >
        username1
      </div>
    </div>
  </div>
</div>
`;
